### ISSUE TITLE
kubectl command to list available resources

### ISSUE THREAD

**Issue by xilabao** [kubernetes/kubernetes#42932]
As the RBAC role need to be related to resources. I think we could use a command to get the supported resources.

ref to https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/admin/authorization.md#referring-to-resources

**Comment by liggitt** [comment]
@kubernetes/sig-cli-feature-requests

**Comment by davidopp** [comment]
This would be cool.

**Comment by smarterclayton** [comment]
I thought we added a command that listed resources in the discover doc a long time ago.

**Comment by smarterclayton** [comment]
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/apiversions.go is just the versions, but principle is the same.  We want to list all the types, their resources, and tie that back to "explain".

**Comment by jimmycuadra** [comment]
What's the best way to determine this information currently? I'm having difficulty figuring out the exact set of permissions (and how they would be written in an RBAC role) for any given kubectl command. The documentation for this right now is split across multiple places and certain things are not explained at all.

Specifically, it's difficult to determine which "API verb" is used for which operations. Take `kubectl proxy` as an example. It's unclear if the way to allow this is to use the "proxy" subresource of a pod or service with "get"/"create"/etc. verbs, or to use the "proxy" verb with the resource (pod, service) directly. `kubectl port-forward` is a similar case.

The [API reference documentation](https://kubernetes.io/docs/api-reference/v1.6/) _almost_ works for this purpose, but it's still difficult. Operations for a given resource are grouped into categories like write, read, status, and proxy, but these things don't obviously correlate with API verbs. Again, proxy is especially confusing as there is no mention of a proxy "verb". The API operations are only shown using their HTTP verbs.

One thing that could help on this document would be a section on the right column under "kubectl command" that shows exactly the rule needed in an RBAC role to allow this command via kubectl. And even then, that assumes an exact 1-to-1 mapping between API endpoints and kubectl commands, which I doubt is always the case. (For example, port fowarding to a pod seems to require both a GET on the pod resource and a CREATE on a port forward subresource.) There also needs to be better documentation about how this information could have been determined in the first place.

I've resorted to manually looking through API requests made with `kubectl --v=8` but even then it's difficult to figure out.

Amusingly, I was the one who wrote https://kubernetes.io/docs/admin/authorization/rbac/#referring-to-resources which is referenced in the original issue description (as well as the table explaining verbs in https://kubernetes.io/docs/admin/authorization/#request-attributes).

tl;dr: It's too difficult right now to go from "I want a user to be able to run `kubectl foo`" to "here is the RBAC rule to use for that" for any case where there isn't a specific example shown in the docs.

**Comment by stewart-yu** [comment]
It seem that relative PR stop updated. If this issue still valuable, i will help fix it @liggitt  @smarterclayton @davidopp
